packages/docusaurus-theme-classic/src/theme/CodeBlock/index.tsx

```diff
@@ -23,154 +28,192 @@ import type {Props} from '@theme/CodeBlock';
+  getTokenProps: GetTokenProps;
+}) {
+  if (line.length === 1 && line[0]!.content === '\n') {
+    line[0]!.content = '';
+  }
+
+  const lineProps = getLineProps({
+    line,
+    ...(showLineNumbers && {className: styles.codeLine}),
+  });
+
+  if (highlight) {
+    lineProps.className += ' docusaurus-highlight-code-line';
+  }
+
+  const lineTokens = line.map((token, key) => (
+    <span key={key} {...getTokenProps({token, key})} />
+  ));
+
+  return (
+    <span {...lineProps}>
+      {showLineNumbers ? (
+        <>
+          <span className={styles.codeLineNumber} />
+          <span className={styles.codeLineContent}>{lineTokens}</span>
+        </>
+      ) : (
+        <>
+          {lineTokens}
+          <br />
+        </>
+      )}
+    </span>
+  );
+}
+
+/**
+ * Best attempt to make the children a plain string so it is copyable. If there
+ * are react elements, we will not be able to copy the content, and it will
+ * return `children` as-is; otherwise, it concatenates the string children
+ * together.
+ */
+function maybeStringifyChildren(children: ReactNode): ReactNode {
+  if (React.Children.toArray(children).some((el) => isValidElement(el))) {
+    return children;
+  }
+  // The children is now guaranteed to be one/more plain strings
+  return Array.isArray(children) ? children.join('') : (children as string);
+}
+
 export default function CodeBlock({
+  children: rawChildren,
+  ...props
+}: Props): JSX.Element {
+  // The Prism theme on SSR is always the default theme but the site theme can
+  // be in a different mode. React hydration doesn't update DOM styles that come
+  // from SSR. Hence force a re-render after mounting to apply the current
+  // relevant styles.
+  const isBrowser = useIsBrowser();
+  const children = maybeStringifyChildren(rawChildren);
+  const CodeBlockComp =
+    typeof children === 'string' ? CodeBlockString : CodeBlockJSX;
+  return (
+    <CodeBlockComp key={String(isBrowser)} {...props}>
+      {children as string}
+    </CodeBlockComp>
+  );
+}
+
+function CodeBlockContainer<T extends 'div' | 'pre'>({
+  as: As,
+  ...props
+}: {as: T} & ComponentProps<T>) {
+  const prismTheme = usePrismTheme();
+  const prismCssVariables = getPrismCssVariables(prismTheme);
+  return (
+    <As
+      // Polymorphic components are hard to type, without `oneOf` generics
+      {...(props as never)}
+      style={prismCssVariables}
+      className={clsx(
+        props.className,
+        styles.codeBlockContainer,
+        ThemeClassNames.common.codeBlock,
+      )}
+    />
+  );
+}
+
+// <pre> tags in markdown map to CodeBlocks. They may contain JSX children.
+// When the children is not a simple string, we just return a styled block
+// without actually highlighting.
+function CodeBlockJSX({children, className}: Props): JSX.Element {
+  return (
+    <CodeBlockContainer
+      as="pre"
+      tabIndex={0}
+      className={clsx(styles.codeBlockStandalone, 'thin-scrollbar', className)}>
```

Comment:
After this refactor I realized the ReactElement version doesn't accept `title`. I tried a bit but because it's inside a `pre` the font is different. Just a note that we _may_ want to support that in the future.